### ISSUE TITLE
Missing stack traces in HTTPChannel.onException

### ISSUE THREAD

**Issue by brettkail-wk** [eclipse/jetty.project#2346]
Similar to #1696, exceptions from `HTTPChannel.onException` are mishandled.  As in that ticket, I also believe this is a bug in `JettyAwareLogger.log` and should be fixed there rather than in the callers: `MessageFormatter.arrayFormat` [removes](https://github.com/qos-ch/slf4j/blob/9b536f4259adf2f01852e86306a3a2d7e4a238c0/slf4j-api/src/main/java/org/slf4j/helpers/MessageFormatter.java#L168-L171) the last argument if it is `Throwable` and does not include the message or stack trace in `getMessage`.  Instead, the `Throwable` is returned via `FormattingTuple.getThrowable`, so `log` should at least be doing:
```
_logger.log(marker, FQCN, level, ft.getMessage(), null, ft.getThrowable());
```

This improves the output to include a stack trace:
```
2018-03-16 14:59:29,450 WARN [qtp2053996178-20] o.e.j.s.HttpChannel [HttpChannel.java:564] /test {}
java.io.IOException: test
	at ...
```

However, you'll see it does include a stray `{}` since the argument.  This could be ignored, or it could be fixed by removing `{}` if present and `getThrowable()` returns non-null.

**Comment by joakime** [comment]
There is no `HTTPChannel.onException()` method in Jetty.

The report in Issue #1696 has nothing to do with JettyAwareLogger, Slf4j, or the Jetty Slf4jLog layer.
It was a fault in the generation/declaration of the logging event itself, way before a logging implementation is even involved.

Looking at your example output ...

> 2018-03-16 14:59:29,450 WARN [qtp2053996178-20] o.e.j.s.HttpChannel [HttpChannel.java:564] /test {}

We can see that it's like 564 in the server implementation of the `HttpChannel`, which looks like this ...

https://github.com/eclipse/jetty.project/blob/e7b411b90d60fb44cd6037ad09e22f7133bdab92/jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java#L569

That will never log an exception stacktrace, on any logger implementation.

The `{} {}` format string is the fault here. 
Each `{}` will take the parameter argument's `.toString()` and replace the `{}` in that location in the format string.

This needs to be fixed, but not in the JettyAwareLogger or MessageFormatter, or Slf4jLog.  It needs to be fixed in the `HttpChannel` codebase itself.

**Comment by brettkail-wk** [comment]
I don't understand what you mean by "That will never log an exception stacktrace, on any logger implementation." since if I remove the `new JettyAwareLogger` from `Slf4jLog` (i.e., logging to slf4j+logback directly), then I do get an exception stack trace.

**Comment by gregw** [comment]
The decision to not log a stack trace was deliberate for these exceptions.  
The code was previously:
```java
        if (failure instanceof BadMessageException | failure instanceof IOException | failure instanceof TimeoutException)
        {
            if (LOG.isDebugEnabled())
                LOG.debug(_request.getRequestURI(), failure);
            else
                LOG.warn("{} {}",_request.getRequestURI(), failure);
        }
```
The reason is that BadMessages, IOExceptions and TimeoutExceptions can be caused by remote clients.    To use the words of a large jetty user I recently spoke to about this: "If we log those stack traces, then users can use our own logging system as a DOS vector against us".

In other words, the cause of the exception is the bad message and the stack trace that throws it is not really relevant.   If needed it is available by debug.
Note also that turning on this stack  traces has resulted in issues like #2543 #2542 #2541 

Thus I'm going to revert this change.

**Comment by brettkail-wk** [comment]
@gregw Yes, I was typing while you were committing :-), thanks.  Still not sure I understand why warning+string is ok but warning+stack is not, but at least better than it was before...

**Comment by gregw** [comment]
@brettkail-wk 
If a user sends a request `GET / HTTP/1.0` we don't log a stack trace saying that we successfully sent them a 200 response.
So if a sends a request `GET / HTTP/99.999` we don't really need to log a stack trace  saying that we successfully sent then a 400 response.

In both cases the server acted correctly and it just happens that we use an exception to exit the handling when we see a bad message.  It is hardly something that should be logged and definitely does not need a stack trace.

**Comment by brettkail-wk** [comment]
@gregw Sure, but in that case it seems we don't need a `warn` at all either?  I suppose the issue here is that we would ideally log fully (with stack trace) exceptions that originate from the handler but not log "internal" or "expected" exceptions.  For client disconnects, I know other servlet containers will throw custom IOException subclasses so that logging can be skipped for those.
